File: back/src/entities/Food.ts
import GameEntity, { GameEntityOptions } from './GameEntity';

interface FoodOptions extends GameEntityOptions {
  value: number;
  id: string;
}

export default class Food extends GameEntity {
  public value: number;
  public lightSize: number;
  public lightDirection: boolean = true;
  public id: string;

  constructor(options: FoodOptions) {
    super(options);
    this.id = options.id;
    this.value = options.value;
    this.lightSize = this.width / 2;
  }

  public action() {
    const lightSpeed = 1;

    this.lightSize += this.lightDirection ? lightSpeed : -lightSpeed;

    // light animate
    if (this.lightSize > this.width || this.lightSize < this.width / 2) {
      this.lightDirection = !this.lightDirection;
    }
  }
}

File: back-go/food.go
package main

import (
	"math"
	"math/rand"

	"github.com/teris-io/shortid"
)

type Food struct {
	GameEntity
	Id             string `msgpack:"id"`
	Value          int
	LightSize      float64
	LightDirection bool
}

func NewFood(x, y float64) *Food {
	return &Food{
		GameEntity: GameEntity{
			Point: Point{
				X: x,
				Y: y,
			},
			Width:  10,
			Height: 10,
		},
		Id:             shortid.MustGenerate(),
		Value:          INITIAL_FOOD_VALUE,
		LightSize:      10,
		LightDirection: true,
	}
}

func NewRandomFood() *Food {
	return NewFood(math.Ceil(rand.Float64()*(MAP_WIDTH-100)+100/2),
		math.Ceil(rand.Float64()*(MAP_HEIGHT-100)+100/2))
}

func (f *Food) Action() {
	lightSpeed := 1.0

	if f.LightDirection {
		f.LightSize += lightSpeed
	} else {
		f.LightSize -= lightSpeed
	}

	if f.LightSize > f.Width || f.LightSize < f.Width/2 {
		f.LightDirection = !f.LightDirection
	}
}
